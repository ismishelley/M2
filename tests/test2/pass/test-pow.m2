void myprint(matrix int [4][4] M) { 
    int i;
    int j;
    for(i = 0; i < M:rows; ++i) {
        for(j = 0; j < M:cols; ++j) {
            printInt(M[i][j]);
        }
        printStr("");
    }
}

void pow(matrix int[4][4] input, int N) {
    matrix int[4][4] res;
    matrix int[4][4] ret;
    int flag;
    flag = 1;
    res = input;
    while(N > 1) {
        if(N - N / 2 * 2 == 0) {
            res = res * res;
            N = N / 2;
        }
        else {
            if(flag == 1) {
                ret = res;
                flag = -1;
            }
            else {
                ret = ret * res;
            }
            res = res * res;
            N = N / 2;
        }
    }
    if(flag == 1) {
        myprint(res);
    }
    else {
        ret = ret * res;
        myprint(ret);
    }
    
}

int main() {
    matrix int [4][4] m;
    int time;
    time = 2;
    m = [[2,9,1,1];
        [0,2,1,1];
        [1,1,1,1];
        [3,1,1,1]
    ];
    pow(m, 6);
    return 0;
}