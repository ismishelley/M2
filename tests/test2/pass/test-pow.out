13984	39720	13788	13788	
4272	11848	4176	4176	
5960	16604	5760	5760	
10744	30196	10360	10360	
